Add plugin support on Windows

julianh-siemens opened the pull request:
This pull request introduces plugin support for Windows, mirroring the existing functionality available on Linux.

**Enabling Plugin Support**:
To enable plugin support during CMake configuration, use the following flag: `-DHWLOC_ENABLE_PLUGINS=true`

**Supported Plugins**:
Currently supported plugins include:

* libxml2 (for XML output/input)
* cuda (for NVIDIA GPUs)
* opencl (for AMD and Intel GPUs)

**Plugin Installation and Configuration**:
Plugins are installed to `.../lib/hwloc/` by default, consistent with the Linux installation path. To ensure plugins are discovered and loaded, this path must be added to the `HWLOC_PLUGINS_PATH` environment variable.

**Demonstration Examples**:
The following examples illustrate successful plugin loading and hardware detection on various Windows systems. In these examples, the libxml2 plugin is shown failing to load (Failed to load plugin: The specified module could not be found.) to demonstrate the behavior when a plugin is built but its path is not correctly configured or available to `HWLOC_PLUGINS_PATH`.

* NVIDIA GPU (via `cuda` plugin):
```
hwloc: Starting plugin dlforeach in C:\Users\<omitted>\hwloc\lib\hwloc
hwloc: Looking under C:\Users\<omitted>\hwloc\lib\hwloc
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_cuda'
hwloc: Plugin contains expected symbol `hwloc_cuda_component'
hwloc: Plugin descriptor `hwloc_cuda' ready
hwloc: Plugin descriptor `hwloc_cuda' queued
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_opencl'
hwloc: Plugin contains expected symbol `hwloc_opencl_component'
hwloc: Plugin descriptor `hwloc_opencl' ready
hwloc: Plugin descriptor `hwloc_opencl' queued
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_xml_libxml'
hwloc: Failed to load plugin: The specified module could not be found.
...
    <object type="OSDev" gp_index="200" name="cuda0" subtype="CUDA" osdev_type="5">
      <info name="Backend" value="CUDA"/>
      <info name="GPUVendor" value="NVIDIA Corporation"/>
      <info name="GPUModel" value="Quadro RTX 5000"/>
      <info name="CUDAGlobalMemorySize" value="16597632"/>
      <info name="CUDAL2CacheSize" value="4096"/>
      <info name="CUDAMultiProcessors" value="48"/>
      <info name="CUDASharedMemorySizePerMP" value="48"/>
    </object>
    <object type="OSDev" gp_index="201" name="opencl0d0" subtype="OpenCL" osdev_type="5">
      <info name="Backend" value="OpenCL"/>
      <info name="OpenCLDeviceType" value="GPU"/>
      <info name="GPUVendor" value="NVIDIA Corporation"/>
      <info name="GPUModel" value="Quadro RTX 5000"/>
      <info name="OpenCLPlatformIndex" value="0"/>
      <info name="OpenCLPlatformName" value="NVIDIA CUDA"/>
      <info name="OpenCLPlatformDeviceIndex" value="0"/>
      <info name="OpenCLComputeUnits" value="48"/>
      <info name="OpenCLGlobalMemorySize" value="16597632"/>
    </object>
```

* AMD GPU (via `opencl` plugin):
```
hwloc: Starting plugin dlforeach in C:\Users\<omitted>\hwloc\lib\hwloc
hwloc: Looking under C:\Users\<omitted>\hwloc\lib\hwloc
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_cuda'
hwloc: Failed to load plugin: The specified module could not be found.
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_opencl'
hwloc: Plugin contains expected symbol `hwloc_opencl_component'
hwloc: Plugin descriptor `hwloc_opencl' ready
hwloc: Plugin descriptor `hwloc_opencl' queued
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_xml_libxml'
hwloc: Failed to load plugin: The specified module could not be found.
...
    <object type="OSDev" gp_index="268" name="opencl0d0" subtype="OpenCL" osdev_type="5">
      <info name="Backend" value="OpenCL"/>
      <info name="OpenCLDeviceType" value="GPU"/>
      <info name="GPUVendor" value="Advanced Micro Devices, Inc."/>
      <info name="GPUModel" value="AMD Radeon PRO W6800"/>
      <info name="OpenCLPlatformIndex" value="0"/>
      <info name="OpenCLPlatformName" value="AMD Accelerated Parallel Processing"/>
      <info name="OpenCLPlatformDeviceIndex" value="0"/>
      <info name="OpenCLComputeUnits" value="30"/>
      <info name="OpenCLGlobalMemorySize" value="31440896"/>
    </object>
```
* Intel GPU (via `opencl` plugin):
```
hwloc: Starting plugin dlforeach in C:\Users\<omitted>\hwloc\lib\hwloc
hwloc: Looking under C:\Users\<omitted>\hwloc\lib\hwloc
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_cuda'
hwloc: Failed to load plugin: The specified module could not be found.
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_opencl'
hwloc: Plugin contains expected symbol `hwloc_opencl_component'
hwloc: Plugin descriptor `hwloc_opencl' ready
hwloc: Plugin descriptor `hwloc_opencl' queued
hwloc: Plugin dlforeach found `C:\Users\<omitted>\hwloc\lib\hwloc\hwloc_xml_libxml'
hwloc: Failed to load plugin: The specified module could not be found.
...
    <object type="OSDev" gp_index="60" name="opencl0d0" subtype="OpenCL" osdev_type="5">
      <info name="Backend" value="OpenCL"/>
      <info name="OpenCLDeviceType" value="GPU"/>
      <info name="GPUVendor" value="Intel(R) Corporation"/>
      <info name="GPUModel" value="Intel(R) Graphics"/>
      <info name="OpenCLPlatformIndex" value="0"/>
      <info name="OpenCLPlatformName" value="Intel(R) OpenCL Graphics"/>
      <info name="OpenCLPlatformDeviceIndex" value="0"/>
      <info name="OpenCLComputeUnits" value="64"/>
      <info name="OpenCLGlobalMemorySize" value="17212992"/>
    </object>
```

julianh-siemens commented:
No worries, there is no rush here.
I only implemented this for CMake, because this is the only way I know how to build hwloc on Windows. With the right conditionals it should also be able to be usable with other build systems.

Our use case is, that we currently have only one hwloc build on Linux for applications that support GPGPU computation (and provide the necessary libraries) and also for applications without GPGPU support/libraries. On Windows we are currently forced to build many different hwloc flavors (no-GPGPU, CUDA, OPENCL). This adds quite a bit of maintenance overhead and duplicated code in shared codebases. Plugin support on Windows would create feature parity to Linux and allow us to use only one hwloc build for all different applications.

bgoglin commented:
> Plugins are installed to `.../lib/hwloc/` by default, consistent with the Linux installation path. To ensure plugins are discovered and loaded, this path must be added to the `HWLOC_PLUGINS_PATH` environment variable.

Does this mean that the end-user will have to set the variable after every installation because the install path is rarely the same on Windows?

bgoglin commented:
Hmmm I am testing right now, after cmake --install, hwloc is looking for plugins in build\lib\hwloc instead of install\lib\hwloc. Setting the env var makes it find the plugins but the default value seems to be wrong?

bgoglin commented on contrib/windows-cmake/CMakeLists.txt:
```diff
@@ -100,6 +101,25 @@ if(HWLOC_WITH_CUDA)
+set(HWLOC_ENABLED_PLUGINS_LIST)
+if(HWLOC_ENABLE_PLUGINS)
+    set(HWLOC_HAVE_PLUGINS 1)
+    set(HWLOC_PLUGINS_PATH ${CMAKE_CURRENT_BINARY_DIR}/lib/hwloc)
```
> @bgoglin is it possible that `${CMAKE_CURRENT_BINARY_DIR}` is not correct here and it points to the build time binary dir and not the install time binary dir?

I am far from an expert on CMake, I could have misconfigured/built things. I already had to change things because my setup wasn't for MSVC (usually I didn't pass --config but rather in a config variable). What I did is:
```
cmake --install-prefix=%cd%/install -D<stuff> -B build
cmake --build build --config Release
cmake --install build --config Release
```

By the way, after overring the plugins path, I get "failed to find component symbol hwloc_cuda_component". I ran dumpbin /symbols on hwloc_cuda.dll and it finds nothing.

bgoglin replied:
Ok so we need to find out why my loading of plugin fails with "failed to find component symbol hwloc_cuda_component"

julianh-siemens commented on contrib/windows-cmake/CMakeLists.txt:
```diff
@@ -148,6 +168,40 @@ target_include_directories(hwloc PUBLIC
+    endif()
+
+    # common settings for all plugins
+    target_compile_definitions(hwloc_${plugin} PRIVATE "HWLOC_INSIDE_PLUGIN")
```
@bgoglin can you check in your local build if this variable is set and triggers `hwloc/topology-cuda.c:190`? I guess this is what is missing in your case.

bgoglin commented:
I couldn't get the code to work here. First plugins are loaded from the build dir instead of the install dir (I worked around the issue by setting the right path in the environment variable). Then loading plugins fails with "failed to find component symbol hwloc_cuda_component" (only tried with Cuda). I don't know how to debug this further.

trivialfis commented:
Thank you for sharing. I will look into this after the python package is ready for initial review.

bgoglin commented:
> Hm, I was hoping that I could reuse the CMake for a normal Linux build. But it looks like there are many configurations baked into the autotool config.

Yes we support lots of Unix systems with lots of libc/features/etc that were available or not in the past. Porting all this to CMake would require a lot of work. CMake is really only for Windows.

bgoglin commented:
How did you add the CUDA library path to your path?

julianh-siemens commented:
For powershell:
```
PS C:\> $env:PATH = "C:\path\to\cuda\12.2.2\win64\bin;" + $env:PATH
```

bgoglin commented:
I am using CMD but I think my path is correct then. I am building with Visual Studio 2022. The hwloc_cuda dependencies are still as listed on https://github.com/open-mpi/hwloc/pull/737#discussion_r2334109988
cudart64_12.dll in my PATH, and I load vcvars64.bat for Visual Studio too. LoadLibraryExA() succeeds, but GetProcAddress() fails with "the specified procedure could not be found".

Adding #define _USRDLL to private/config.h helps, I now see the hwloc_cuda_component in dumpbin /exports. Loading the plugin succeeds but fails later because it couldn't find hwloc_backend_alloc() in the rest of the process, so I guess the core symbols aren't exported from libhwloc, but I don't have a libhwloc.dll to check. However I also see all hwloc core symbols inside hwloc_cuda.dll (related to #738?) so there's also an issue with hwloc_cuda.dll embedding all the core library while it should not (otherwise some global variables will be duplicated in the plugin and core lib). I think we cannot have a static lib in the plugin case.

julianh-siemens commented:
Rebased to current master to resolve merge conflicts and fixed the hard coded plugin path to point to the install location instead of the build location.

bgoglin commented:
-DBUILD_SHARED_LIBS:BOOL=ON fixes the missing hwloc_cuda_component missing symbol, no need for _USRDLL, thanks. But now it fails later with "hwloc_backend_alloc()". hwloc_cuda.dll still exports all hwloc symbols according to dumpbin /exports (and hwloc_cuda.dll is twice bigger than hwloc.dll which makes no sense) while hwloc.dll doesn't export anything (which might explains why it doesn't find hwloc_backend_alloc). So something is embedded hwloc core symbols inside the plugin and exporting them while they aren't exported from the main hwloc.dll

bgoglin commented:
Ok it works now. I took your last code and built/loaded the cuda plugin in both cmd (as earlier) and powershell (new). Not even need to put the cuda path in the cmake flags (it was in my PATH). The plugin is loaded and GPU is visible in lstopo. And I can confirm that reconfiguring inside a directory previously built without SHARED_LIBS causes the issue. I am not familiar with CMake but I don't like that behavior :)

Could you:
* Enable BUILD_SHARED_LIBS when PLUGINS are enabled. By the way, if you understand the difference between BUILD_SHARED_LIBS and HWLOC_BUILD_SHARED_LIBS, I'd like to merge them. But maybe the former is standard while the latter is hwloc-specific and we need to keep both?
* rename your commit title into something like "Add plugin support in Windows CMake builds"
* add a sentence in your commit message saying that building plugins is only added in contrib/windows-cmake but the support for loading them may work for non-cmake builds (not tested yet)

julianh-siemens commented:
Great that we finally managed to iron this out.

I adjusted the commit message like you suggested.

Regarding BUILD_SHARED_LIBS and HWLOC_BUILD_SHARED_LIBS:
The former is CMake standard, so I deprecated the custom option. There will be an info message whenever it is set by the user:
```
-- The option HWLOC_BUILD_SHARED_LIBS has been removed. Please use BUILD_SHARED_LIBS instead.
```

Let me know if this should be a fatal error.

Regarding the handling of BUILD_SHARED_LIBS and whether it is (not) set to ON/OFF **in combination with plugin support**, I have opted for the following:
* If it is not set by the user, it will be forced to ON and an info message will be printed:
```
-- HWLOC plugin support requires BUILD_SHARED_LIBS to be enabled, but it was not set. Setting BUILD_SHARED_LIBS to ON
```
* If it was set to ON by the user, no extra output or action is taken.
* If it was set to OFF by the user, an fatal error message is printed:
```
CMake Error at CMakeLists.txt:122 (message):
  HWLOC plugin support requires BUILD_SHARED_LIBS to be enabled, but BUILD_SHARED_LIBS is set to false.
```
